# L2 Escape Model

Consequent Reversibility was not included in the final escape models because, in this subgroup structure, it is partially encoded by Goal_Type and produced unstable estimates when entered together with Goal_Type.

Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Overall Subgroup:
1. Create escape_L2
2. Counts and Proportions
3. Logistic Models
4. Model Comparison

Read data

In [5]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create escape_2

In [7]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

2. Counts and Proportions

In [ ]:
escape_L2_counts = pd.crosstab(
    subgroup_theoretical["escape_L2"],
    subgroup_theoretical["Response_Full"])

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct
escape_L2,,,
0,0,38,0
1,96,0,61


In [12]:
escape_L2_props = subgroup_theoretical["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.808081
0    0.191919
Name: proportion, dtype: float64

Counts and Proportions by Condition

In [16]:
counts_condition = pd.crosstab(
    [subgroup_theoretical["Goal_Type"], subgroup_theoretical["Agent"], subgroup_theoretical["Consequent_Reversibility"]],
    subgroup_theoretical["escape_L2"],
    margins = True)

counts_condition

escape_L2                                          0    1  All
Goal_Type         Agent Consequent_Reversibility              
goal_frequent     0     Yes                       14   28   42
                  1     Yes                        2   22   24
goal_non_frequent 0     No                         9   33   42
                  1     No                         1   23   24
no_goal           0     Yes                        8   34   42
                  1     Yes                        4   20   24
All                                               38  160  198

In [14]:
props_condition = pd.crosstab(
    [subgroup_theoretical["Goal_Type"], subgroup_theoretical["Agent"], subgroup_theoretical["Consequent_Reversibility"]],
    subgroup_theoretical["escape_L2"],
    normalize= "index")

props_condition

escape_L2                                                0         1
Goal_Type         Agent Consequent_Reversibility                    
goal_frequent     0     Yes                       0.333333  0.666667
                  1     Yes                       0.083333  0.916667
goal_non_frequent 0     No                        0.214286  0.785714
                  1     No                        0.041667  0.958333
no_goal           0     Yes                       0.190476  0.809524
                  1     Yes                       0.166667  0.833333

Agency appears to help participants escape L2. 

The condition with the strongest L1 transfer (goal_non_frequent, agent=1) also exhibits the highest L2 escape rate (95.8%). This suggests that failure does not occur at the L2 suppression stage, but at the subsequent resolution stage.

In addition, if there is a goal representation available, agency seems to help participants commit by leaving the L2 space; however, if there is no goal representation, agency seems to have much less to work with. 

3. Logistic Models

Import Libraries

In [20]:
import statsmodels.formula.api as smf
from scipy.stats import chi2

Binary Outcome

In [21]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

In [22]:
subgroup_theoretical["escape_L2"].value_counts()

escape_L2
1    160
0     38
Name: count, dtype: int64

**Model E1**
escape_L2 ~ Goal_Type

In [48]:
E1 = smf.logit(
    "escape_L2 ~ Goal_Type",
    data=subgroup_theoretical
    ).fit()

print(E1.summary())

Optimization terminated successfully.
         Current function value: 0.484442
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      195
Method:                           MLE   Df Model:                            2
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                0.009309
Time:                        17:12:15   Log-Likelihood:                -95.920
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                    0.4060
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          1.1394      0.287      3.967     

**Model E2**
escape_L2 ~ Goal_Type + Agency

In [51]:
E2_2 = smf.logit(
    "escape_L2 ~ Consequent_Reversibility + Agent",
    data= subgroup_theoretical
    ).fit()

print(E2_2.summary())

Optimization terminated successfully.
         Current function value: 0.468220
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      195
Method:                           MLE   Df Model:                            2
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.04248
Time:                        17:13:44   Log-Likelihood:                -92.708
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.01636
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                           1.4127      0.359      3.932  

In [50]:
E2 = smf.logit(
    "escape_L2 ~ Goal_Type + Agent",
    data= subgroup_theoretical
    ).fit()

print(E2.summary())

Optimization terminated successfully.
         Current function value: 0.466314
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      194
Method:                           MLE   Df Model:                            3
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.04638
Time:                        17:12:44   Log-Likelihood:                -92.330
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.02954
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.8081      0.311      2.599     

**Model E3**
escape_L2 ~ Consequent_Reversibility + Goal_Type + Agent

In [28]:
E3 = smf.logit(
    "escape_L2 ~ Goal_Type + Agent",
    data=subgroup_theoretical
    ).fit()

print(E3.summary())

Optimization terminated successfully.
         Current function value: 0.466314
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      194
Method:                           MLE   Df Model:                            3
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.04638
Time:                        16:21:24   Log-Likelihood:                -92.330
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.02954
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.8081      0.311      2.599     

**Model E4**
escape_L2 ~ Goal_Type x Agent

In [30]:
E4 = smf.logit(
    "escape_L2 ~ Goal_Type * Agent",
    data=subgroup_theoretical
).fit()

print(E4.summary())

Optimization terminated successfully.
         Current function value: 0.458893
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      192
Method:                           MLE   Df Model:                            5
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.06156
Time:                        16:28:55   Log-Likelihood:                -90.861
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.03590
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.6931      0.3

In [52]:
E4_2 = smf.logit(
    "escape_L2 ~ Consequent_Reversibility * Agent",
    data=subgroup_theoretical
).fit()

print(E4_2.summary())

Optimization terminated successfully.
         Current function value: 0.466503
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      194
Method:                           MLE   Df Model:                            3
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.04599
Time:                        17:15:32   Log-Likelihood:                -92.368
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.03056
                                            coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------
Intercept                                 1.2993      

4. Model Comparisons

In [32]:
from scipy.stats import chi2

def lr_test(model_small, model_large):

    LR = 2 * (model_large.llf - model_small.llf)

    df = model_large.df_model - model_small.df_model

    p = chi2.sf(LR, df)

    return pd.Series({
        "LR": LR,
        "df": df,
        "p_value": p
    })

In [33]:
lr_test(E3, E4)

LR         2.938946
df         2.000000
p_value    0.230047
dtype: float64

## Actors Subgroup:
1. Create escape_L2
2. Counts and Proportions
3. Logistic Models
4. Model Comparison

Read data

In [34]:
subgroup_actors = subgroup_theoretical[subgroup_theoretical["Focus"] == "I"].copy()

1. Create escape_L2

In [35]:
subgroup_actors["escape_L2"] = (
    subgroup_actors["Response_Full"] != "L2_other"
).astype(int)

2. Counts and Proportions

In [36]:
escape_L2_counts = pd.crosstab(
    subgroup_actors["escape_L2"],
    subgroup_actors["Response_Full"])

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct
escape_L2,,,
0,0,24,0
1,53,0,40


In [37]:
escape_L2_props = subgroup_actors["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.8
0    0.2
Name: proportion, dtype: float64

Counts and Proportions by Condition

In [38]:
counts_condition = pd.crosstab(
    [subgroup_actors["Goal_Type"], subgroup_actors["Agent"], subgroup_actors["Consequent_Reversibility"]],
    subgroup_actors["escape_L2"],
    margins = True)

counts_condition

escape_L2                                          0   1  All
Goal_Type         Agent Consequent_Reversibility             
goal_frequent     0     Yes                        8  16   24
                  1     Yes                        2  14   16
goal_non_frequent 0     No                         8  16   24
                  1     No                         1  15   16
no_goal           0     Yes                        3  21   24
                  1     Yes                        2  14   16
All                                               24  96  120

In [39]:
props_condition = pd.crosstab(
    [subgroup_actors["Goal_Type"], subgroup_actors["Agent"], subgroup_actors["Consequent_Reversibility"]],
    subgroup_actors["escape_L2"],
    normalize= "index")

props_condition

escape_L2                                                0         1
Goal_Type         Agent Consequent_Reversibility                    
goal_frequent     0     Yes                       0.333333  0.666667
                  1     Yes                       0.125000  0.875000
goal_non_frequent 0     No                        0.333333  0.666667
                  1     No                        0.062500  0.937500
no_goal           0     Yes                       0.125000  0.875000
                  1     Yes                       0.125000  0.875000

3. Logistic Models

Binary Outcome

In [40]:
subgroup_actors["escape_L2"] = (
    subgroup_actors["Response_Full"] != "L2_other"
).astype(int)

In [41]:
subgroup_actors["escape_L2"].value_counts()

escape_L2
1    96
0    24
Name: count, dtype: int64

4. Model comparisons

**E1 Model**: 
escape_L2 ~ Consequent_Reversibility

In [60]:
E1 = smf.logit(
    "escape_L2 ~ Goal_Type",
    data=subgroup_actors
    ).fit()

print(E1.summary())

Optimization terminated successfully.
         Current function value: 0.490756
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      117
Method:                           MLE   Df Model:                            2
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.01928
Time:                        17:24:48   Log-Likelihood:                -58.891
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                    0.3143
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          1.0986      0.365      3.009     

For actors, consequent reversibility did not significantly predict L2 escape (p = .629), suggesting that reversibility alone is insufficient to trigger abandonment of L2 interpretations. This supports the view that reversibility functions as a source of latent cognitive load whose effects become observable only when additional representational demands are introduced.

In [61]:
E1_2 = smf.logit(
    "escape_L2 ~ Consequent_Reversibility",
    data=subgroup_actors
    ).fit()

print(E1_2.summary())

Optimization terminated successfully.
         Current function value: 0.499440
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      118
Method:                           MLE   Df Model:                            1
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                0.001924
Time:                        17:24:50   Log-Likelihood:                -59.933
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                    0.6307
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                           1.2368      0.379      3.266  

**E2 Model**: 
escape_L2 ~  Goal_Type + Agent

In [54]:
E2 = smf.logit(
    "escape_L2 ~ Goal_Type + Agent",
    data= subgroup_actors
    ).fit()

print(E2.summary())

Optimization terminated successfully.
         Current function value: 0.469898
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      116
Method:                           MLE   Df Model:                            3
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.06096
Time:                        17:18:39   Log-Likelihood:                -56.388
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.06234
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.7203      0.400      1.799     

In [ ]:
E2_2 = smf.logit(
    "escape_L2 ~ Consequent_Reversibility + Agent",
    data= subgroup_actors
    ).fit()

print(E2_2.summary())

Optimization terminated successfully.
         Current function value: 0.478907
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      117
Method:                           MLE   Df Model:                            2
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.04296
Time:                        17:20:09   Log-Likelihood:                -57.469
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.07581
                                      coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                           0.8695      0.410      2.118  

**E3 Model**
escape_L2 ~ Goal_Type * Agent

In [56]:
E3 = smf.logit(
    "escape_L2 ~ Goal_Type * Agent",
    data= subgroup_actors
    ).fit()

print(E3.summary())

Optimization terminated successfully.
         Current function value: 0.461604
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      114
Method:                           MLE   Df Model:                            5
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.07753
Time:                        17:21:04   Log-Likelihood:                -55.392
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.09726
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.6931      0.4

In [58]:
E3_2 = smf.logit(
    "escape_L2 ~ Consequent_Reversibility * Agent",
    data= subgroup_actors
    ).fit()

print(E3_2.summary())

Optimization terminated successfully.
         Current function value: 0.474254
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      116
Method:                           MLE   Df Model:                            3
Date:                Mon, 15 Jun 2026   Pseudo R-squ.:                 0.05225
Time:                        17:22:21   Log-Likelihood:                -56.910
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.09894
                                            coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------
Intercept                                 0.6931      

In [62]:
lr_test(E1, E2)

LR         2.843966
df         0.000000
p_value         NaN
dtype: float64

In [63]:
lr_test(E2, E3)

LR         4.152614
df         3.000000
p_value    0.245449
dtype: float64

# Short Comparison:
1. Does the same predictor structure hold?
2. Does effect size/ model fit improve?